tools/src/wyvern/tools/typedAST/core/declarations/DefDeclaration.java

```diff
@@ -235,29 +254,82 @@ public Environment extendName(Environment env, Environment against) {
 	@Override
 	public DeclType genILType(GenContext ctx) {
 		List<FormalArg> args = new LinkedList<FormalArg>();
-		for (NameBinding b : argNames) {
+
+        ctx = this.serializeArguments(args, ctx);
+
+		DefDeclType ret = new DefDeclType(getName(), getResultILType(ctx), args);
+		return ret;
+	}
+
+    private GenContext serializeArguments(List<FormalArg> args, GenContext ctx) {
+        if(isGeneric()) {
+            ValueType type = this.genericStructuralType();
+            args.add(new FormalArg(GENERIC_PREFACE, type));
+            ctx = ctx.extend(GENERIC_PREFACE, new Variable(GENERIC_PREFACE), type);
+        }
+
+        for (NameBinding b : argNames) {
 			String bName = b.getName();
-			ValueType type = b.getType().getILType(ctx);
-			FormalArg fa = new FormalArg(bName, type);
+            Type t =  b.getType();
+            ValueType type = null;
+            if(
```

Comment:
Don't need this if statement if you use TypeGenContext above.  The else clause will just work.